### ISSUE TITLE
Using typeof(...) does not narrow union types

### ISSUE THREAD

**Issue by jdmichal** [Microsoft/TypeScript#9391]
**TypeScript Version:**  1.8.5

**Code**

[TypeScript Playground](http://www.typescriptlang.org/play/#src=interface%20SomeInterface%20{%0D%0A%09label%20%3A%20string%3B%0D%0A%09value%20%3A%20any%3B%0D%0A}%0D%0A%0D%0Afunction%20doStuff%28thing%20%3A%20string%20|%20SomeInterface%29%20%3A%20string%20{%0D%0A%09if%20%28typeof%28thing%29%20%3D%3D%3D%20%22string%22%29%20{%0D%0A%09%09return%20thing%3B%0D%0A%09}%20else%20{%0D%0A%09%09return%20thing.x%3B%0D%0A%09}%0D%0A}%0D%0A)

```ts
interface SomeInterface {
    x : string;
}

function doStuff(thing : string | SomeInterface) : string {
    if (typeof(thing) === "string") {
        return thing;
    } else {
        return thing.x;
    }
}
```

**Expected behavior:**

`thing` would have type `string` in the then-clause, and `SomeInterface` in the else-clause.

**Actual behavior:** 

`thing` remains the union-type `string | SomeInterface` in both clauses.

**Workaround:**

Removing the parenthesis from the `typeof` call will cause the types to narrow as expected.

**Comment by kitsonk** [comment]
[Remove the parentheses](http://www.typescriptlang.org/play/#src=interface%20SomeInterface%20%7B%0D%0A%09label%20%3A%20string%3B%0D%0A%09value%20%3A%20any%3B%0D%0A%7D%0D%0A%0D%0Afunction%20doStuff(thing%20%3A%20string%20%7C%20SomeInterface)%20%3A%20string%20%7B%0D%0A%09if%20(typeof%20thing%20%3D%3D%3D%20%22string%22)%20%7B%0D%0A%09%09return%20thing%3B%0D%0A%09%7D%20else%20%7B%0D%0A%09%09return%20thing.x%3B%0D%0A%09%7D%0D%0A%7D%0D%0A)...  There are limits to the effort TypeScript will go through in trying to narrow types.  Ryan commented on this before, though I can't remember the ticket.  `typeof` is not a function, it is an operator, so the `()` are wholly unnecessary:

```typescript
interface SomeInterface {
	label : string;
	value : any;
}

function doStuff(thing : string | SomeInterface) : string {
	if (typeof thing === "string") {
		return thing;
	} else {
		return thing.x;
	}
}
```

**Comment by RyanCavanaugh** [comment]
I think we should probably start unwrapping one level of parens here since `typeof(x)` is probably pretty common. Most places this happens anyway and type guards are smart enough now it'd be very rare to want to "turn them off".

**Comment by jdmichal** [comment]
@kitsonk That's the workaround I listed. The parens are unnecessary, but they are a not-uncommon coding style. Up until about 10 days ago, I simply thought that the narrowing wasn't working. I actually thought that TypeScript couldn't narrow an interface, since I could theoretically add attributes to any object such that it matches the interface. But my only issue was using parens; how silly!

At the least, I would consider some warnings in the documentation about parens not working as sufficient to close this ticket. Though I would obviously prefer the code I posted to work as expected.

**Comment by kitsonk** [comment]
Actually @RyanCavanaugh commented on the limits of TypeScript unwrapping things, so him suggesting they get unwrapped is more than good enough for me to close my mouth now. 😄

**Comment by martine** [comment]
Perhaps some layer (the compiler? or tslint?) could warn on `typeof(x)` expressions to avoid this problem.

**Comment by jdmichal** [comment]
I'd be happy with pretty much anything other than the current status of silently failing to narrow types with no documentation or warning indicating that the statement could be changed to enable narrowing. A warning would be fine, though I would definitely want it in the compiler and not a separate linter. But this seems like one of those instances where by the time the compiler could issue a warning, it could instead just unwrap the statement.

For reference, [here's the documentation page](https://www.typescriptlang.org/docs/handbook/advanced-types.html) that describes type guards, including `typeof`. No mention of parens breaking the guarding behavior. A warning callout there would be nice, if it's decided against the compiler unwrapping.

**Comment by jdmichal** [comment]
Looking at the code, it *looks* like this might be addressed in 2.0.0? Specifically, [this commit](https://github.com/Microsoft/TypeScript/commit/4a8f94a5535548a3b9cc8eaf4c02c0830fba088f) seems to add parens unwrapping via the added [isNarrowableOperand](https://github.com/Microsoft/TypeScript/commit/4a8f94a5535548a3b9cc8eaf4c02c0830fba088f#diff-7455dc48a52c4974a3e3671fcfb1dfb9R638) function. I don't have 2.0.0-beta available; could someone confirm?

**Comment by RyanCavanaugh** [comment]
Compiles without error in `master` branch

**Comment by jdmichal** [comment]
If so, looks like this can be closed once 2.0.0 releases then.
